Sources/SwiftGRPCNIO/HTTP1ToRawGRPCClientCodec.swift

```diff
@@ -114,7 +116,8 @@ extension HTTP1ToRawGRPCClientCodec: ChannelInboundHandler {
       throw GRPCError.client(.invalidState("received body while in state \(state)"))
     }
 
-    for message in try self.messageReader.consume(messageBuffer: &messageBuffer, compression: inboundCompression) {
+    self.messageReader.append(buffer: &messageBuffer)
+    while let message = try self.messageReader.nextMessage() {
```

Comment:
this really feels like this should be done as a `ByteToMessageDecoder`. I think my re-entrancy concerns are real here and I guess you might want to add a test case: Imagine this happens:

- you enter `channelRead` -> `processBody`
- let's assume `self.messageReader` contains 3 messages
- the first of the 3 messages is sent through the pipeline using the `ctx.fireChannelRead(self.wrapInboundOut(.message(message)))` below
- another `ChannelHandler` further down the pipeline decides to `channel.close`
- `channel.close` will force NIO to drain the socket buffer, ie. there might be more input coming from the socket
- the remaining bytes in the socket buffer are now fed back into `channelRead` but we're still on the same stack frame

what you'll now get is reordering. So let's assume there were 4 messages sent in total. As in the example above, 3 were delivered in one `channelRead`. When delivering the 1st, the next `ChannelHandler` will see this but as above, that next `ChannelHandler` will now cause a `close` to happen which might trigger another `channelRead` where we'd receive message 4 but that's now processed directly.

So a subsequent `ChannelHandler` would see `1, 4, 2, 3`

Reply:
here's one of our test cases that fakes this scenario: https://github.com/apple/swift-nio/blob/87ee03c4ce4eb161519aff9b9dba65ed3f2ae23a/Tests/NIOTests/CodecTest.swift#L529-L564

in this test case instead of triggering the complicated case where `channelRead -> close -> channelRead` it just uses the `EmbeddedChannel` and fabricates and inbound messages whilst in `decode` which leads to exactly the same situation.

It's also not just `close` within `channelRead` that can cause re-entrancy. With the SSL handler I believe `write` can trigger a `channelRead`

Reply:
@glbrntt / @MrMage what I just realise here is that unfortunately you would need a `MessageToMessageDecoder` here and not a `ByteToMessageDecoder` (because your `InboundIn == HTTPClientResponsePart` and not `ByteBuffer`). Unfortunately, NIO hasn't implemented a `MessageToMessageDecoder` yet... We should do that soon.

Reply:
That makes sense, thanks for the detailed explanation @weissi!

Is `MessageToMessageDecoder` planned work? Do you have any suggestions for mitigating this in the meantime?

Reply:
@glbrntt we haven't started work on it but this could be a good argument for doing it :). The mitigation is annoying but not too hard: You need to remember if you're currently calling out and if you are, you need to buffer any further arriving data... Something like

```swift
enum State {
    case normal
    case alreadyCallingOut(CircularBuffer<InboundIn>?)
}
```

then if `normal`, you'd just do everything as before but before calling out you would just enqueue the thing into the CircularBuffer for consumption when the `fireChannelRead` returns.

It's quite an annoying bit of code to write though. You could also decide to file a bug and fix this later. Please feel free to file a NIO feature request if you do so.

Reply:
@weissi wow, that re-ordering does sound _bad_. Sorry for the silly question, but how would this be mitigated if we implemented this as a `ChannelHandler` instead?

Reply:
I don't think Implementing it as a `ChannelHandler` _without the above state tracking_ would mitigate this by itself.

`ByteToMessageDecoder` (well, `MessageToMessageDecoder` when it exists) would handle the state tracking for us.

Reply:
indeed, what @glbrntt says. Today, you'd need to implement the buffering on re-entrancy yourself which is a bit annoying.

Reply:
Ouch. Does this mean that any `ChannelHandler` potentially needs to be re-entrant? Wouldn't it be easier to have that handled by the framework, or have the framework simply avoid such re-entrancy? (You have thought more about this than I of course; there are probably good reasons why the mentioned `Decoder`s are the better solution to this.)

Reply:
> Ouch. Does this mean that any `ChannelHandler` potentially needs to be re-entrant? 

Correct, unless you have a `Channel` or `ChannelHandler` that guarantees that won't cause re-entrancy. You can also implement a handler up front that breaks re-entrancy originating from the `Channel` itself. [`HTTPServerPipelineHandler`](https://github.com/apple/swift-nio/blob/master/Sources/NIOHTTP1/HTTPServerPipelineHandler.swift) does things along those lines.

> Wouldn't it be easier to have that handled by the framework, or have the framework simply avoid such re-entrancy? (You have thought more about this than I of course; there are probably good reasons why the mentioned `Decoder`s are the better solution to this.)

It would make things easier but unfortunately also slow. That adds basically two buffers to each channel handler. We are thinking about making a thing which you can opt in to which solves re-entrancy for you. It could for example be a wrapper around any `ChannelHandler` breaking re-entrancy. There's a lot that we can improve here, for now we're not super helpful, just like Netty is. (unless you use `ByteToMessageDecoder`s)

For gRPC it's actually quite interesting because you're using `swift-nio-http2` which might already give you the guarantees you need. To implement HTTP/2 @lukasa already had to deal with reentrancy being tough a lot so it might be that what he implemented gives you no re-entrancy for handlers further back anyway and we can ignore all this here :).

---

side note: When I say a ChannelHandler could guarantee no re-entrancy to the back of it I only mean re-entrancy caused by the Channel or handlers before. No ChannelHandler will be able to protect about crazy code like

```swift
context.channel.pipeline.context(type: TypeOfHandlerInFrontOfMe.self).flatMap { context in
    context.fireChannelRead(...)
}
```

where a handler might inject data at any point to a handler before it.